snap/snapcraft.yaml

```diff
@@ -24,43 +25,125 @@ layout:
 apps:
   hotsos:
     command: bin/hotsos
+
 parts:
-  hotsos-cli:
-    plugin: dump
-    source: .
-    override-build: |
-      rm -rf tests
-      git rev-parse --short HEAD > repo-info
-      snapcraftctl build
-    organize:
-      defs: etc/hotsos/defs
-    build-packages: [git]
   hotsos:
-    plugin: python
+    plugin: dump
     source: .
+    build-packages:
+      - git
+      - python3-pip
+      - python3-venv
+      - rsync
     stage-packages:
       - bsdmainutils
       - coreutils
       - dpkg
+      - libc6
+      - libpython3.8-minimal
+      - libpython3.8-stdlib
+      - python3.8-minimal
       - systemd
-    python-packages:
-      - cryptography==3.4.8
-      - pip
-      - setuptools
-      - wheel
-    requirements:
-      - requirements.txt
     build-environment:
       - "CRYPTOGRAPHY_DONT_BUILD_RUST": "1"
-    filesets:
+    override-build: |
+      PYTHON=python3.8
+
+      # In case we are debugging this snap and are running the build stage
+      # multiple times (as in `snapcraft build --shell`), we have to guard
+      # against running `venv` multiple times.
+      if [[ ! -x ${SNAPCRAFT_PART_INSTALL}/bin/activate ]]; then
+        /usr/bin/python3 -m venv ${SNAPCRAFT_PART_INSTALL}
+      fi
+
+      # Do not source `activate` so that consecutive rebuilding the snap (for
+      # debugging) works.
+      ${SNAPCRAFT_PART_INSTALL}/bin/pip install wheel
+      ${SNAPCRAFT_PART_INSTALL}/bin/pip install -r requirements.txt
+
+      # Build `hotsos`
+      #
+      # Note that we will need `dist-info` when loading `hotsos` which is not
+      # built with a simple `setup.py install`. Building a `bdist_wheel` and
+      # then installing it via `pip` addresses this shortcoming.
+      ${SNAPCRAFT_PART_INSTALL}/bin/python3 setup.py bdist_wheel
+      ${SNAPCRAFT_PART_INSTALL}/bin/pip install dist/hotsos-*.whl
+
+      # Update repo-info
+      git rev-parse --short HEAD > repo-info
+      # Install repo-info
+      install --mode 0644 repo-info ${SNAPCRAFT_PART_INSTALL}
+
+      # Rewrite the shebang to point to the correct Python interpreter. Only
+      # touch the first line and absorb the whole line. Note that the matching
+      # expression cannot be more specific since the path to the Python
+      # interpreter depends on the build environment. Notably, local builds
+      # (`/root`) differ from builds on launchpad (`/build`).
+      sed --in-place \
+        --expression "1 s:^#\!/.*$:#\!/snap/hotsos/current/usr/bin/${PYTHON}:" \
+        ${SNAPCRAFT_PART_INSTALL}/bin/hotsos
+
+      # Add a missing link in snap to Python interpreter.
+      ln -sf ../usr/bin/${PYTHON} ${SNAPCRAFT_PART_INSTALL}/bin/python3
```

Comment:
Is it okay for the `source` to be a relative path? Would that cause any problems if python3 is invoked from a different working directory?

Reply:
It's relative to `${SNAPCRAFT_PART_INSTALL}/bin`. In other words it will link to
`/snap/hotsos/current/usr/bin` in the snap.